internal/dataplane/sendconfig/dbmode.go

```diff
@@ -14,29 +17,37 @@ import (
 
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/deckerrors"
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/metrics"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/util"
 )
 
 // UpdateStrategyDBMode implements the UpdateStrategy interface. It updates Kong's data-plane
 // configuration using decK's syncer.
 type UpdateStrategyDBMode struct {
-	client      *kong.Client
-	dumpConfig  dump.Config
-	version     semver.Version
-	concurrency int
-	isKonnect   bool
+	client            *kong.Client
+	dumpConfig        dump.Config
+	version           semver.Version
+	concurrency       int
+	isKonnect         bool
+	logger            logr.Logger
+	resourceErrors    []ResourceError
+	resourceErrorLock *sync.Mutex
```

Comment:
I think using `sync.Mutex` (no pointer) here is better since the lock is not shared between some other `UpdateStrategyDBMode`s else.

Reply:
This is at odds with our frequent use of non-pointer receivers for this type:

```
$ go vet ./...                                             
# github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig
# [github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig]
internal/dataplane/sendconfig/dbmode.go:63:9: return copies lock value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
internal/dataplane/sendconfig/dbmode.go:66:9: Update passes lock by value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
internal/dataplane/sendconfig/dbmode.go:192:9: MetricsProtocol passes lock by value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
internal/dataplane/sendconfig/dbmode.go:196:9: Type passes lock by value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
internal/dataplane/sendconfig/dbmode.go:200:9: currentState passes lock by value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
internal/dataplane/sendconfig/dbmode.go:209:9: targetState passes lock by value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
# github.com/kong/kubernetes-ingress-controller/v3/test/kongintegration
# [github.com/kong/kubernetes-ingress-controller/v3/test/kongintegration]
test/kongintegration/translator_golden_tests_outputs_test.go:111:67: call of ensureGoldenTestOutputIsAccepted copies lock value: github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig.UpdateStrategyDBMode contains sync.Mutex
```

Effectively it _is_ being shared with other `UpdateStrategyDBMode` instances, they're just copies of itself :neutral_face:

I don't know why we actually have it that way, but it could potentially result in other unexpected behavior changes if we refactor to pointer receivers. Do you know if there's some other issue with the pointer mutex?

I wouldn't expect any since it's private, and the instances it's shared with are things that should be using the same lock. We _do_ need a pointer receiver for `UpdateStrategyDBMode.HandleEvents()`, else the `ResourceError` instances it collects disappear into the void when the goroutine ends.

Reply:
Given that we are already copying `UpdateStrategyDBMode` through functions, it takes less effect to use pointer to the mutex.